src/Microsoft.Health.Dicom.Core/Features/Query/QueryService.cs

```diff
@@ -64,15 +65,55 @@ public class QueryService : IQueryService
-        IEnumerable<DicomDataset> responseMetadata = instanceMetadata.Select(m => responseBuilder.GenerateResponseDataset(m));
+        IEnumerable<DicomDataset> responseMetadata = null;
+
+        if (QueryLimit.ContainsComputedTag(queryExpression.IELevel, queryExpression.IncludeFields.DicomTags))
+        {
+            var versions = queryResult.DicomInstances.Select(i => i.Version).ToList();
+
+            if (queryExpression.IELevel == Messages.ResourceType.Study)
+            {
+                var studyComputedResults = await _queryStore.GetStudyResultAsync(partitionKey, versions, cancellationToken);
+                var studyResultUidMap = studyComputedResults.ToDictionary<StudyResult, string>(a => a.StudyInstanceUid, StringComparer.OrdinalIgnoreCase);
+                responseMetadata = await Task.WhenAll(queryResult.DicomInstances.Select(x =>
+                GenerateQueryResult(x, studyResultUidMap[x.StudyInstanceUid].DicomDataset, responseBuilder, getFullMetadata: true, cancellationToken)));
+            }
+            else if (queryExpression.IELevel == Messages.ResourceType.Series)
+            {
+                var seriesComputedResults = await _queryStore.GetSeriesResultAsync(partitionKey, versions, cancellationToken);
+                var seriesResultMap = seriesComputedResults.ToDictionary<SeriesResult, string>(a => a.StudyInstanceUid + "-" + a.SeriesInstanceUid, StringComparer.OrdinalIgnoreCase);
+                responseMetadata = await Task.WhenAll(queryResult.DicomInstances.Select(x =>
+                GenerateQueryResult(x, seriesResultMap[x.StudyInstanceUid + "-" + x.SeriesInstanceUid].DicomDataset, responseBuilder, getFullMetadata: true, cancellationToken)));
+            }
+        }
+        if (responseMetadata == null)
+        {
+            IEnumerable<DicomDataset> instanceMetadata = await Task.WhenAll(
+                queryResult.DicomInstances.Select(x => _metadataStore.GetInstanceMetadataAsync(x, cancellationToken)));
+
+            responseMetadata = instanceMetadata.Select(m => responseBuilder.GenerateResponseDataset(m));
+        }
 
         return new QueryResourceResponse(responseMetadata, queryExpression.ErroneousTags);
     }
 
+    // TODO version the API to change the default results to only include indexed columns and control the getFullMetadata param
+    private async Task<DicomDataset> GenerateQueryResult(
+        VersionedInstanceIdentifier versionedInstance,
+        DicomDataset computedMetadata,
+        QueryResponseBuilder queryResponseBuilder,
+        bool getFullMetadata,
+        CancellationToken cancellationToken)
+    {
+        if (getFullMetadata)
+        {
+            DicomDataset fullMetadata = await _metadataStore.GetInstanceMetadataAsync(versionedInstance, cancellationToken);
+            return queryResponseBuilder.GenerateResponseDataset(fullMetadata, computedMetadata);
+        }
+        return queryResponseBuilder.GenerateResponseDataset(computedMetadata);
```

Comment:
In what condition this line executes?

Reply:
Right now no, there is a todo to use it in the future, that is my next PR